AAInstrumentation/unique_hashes.py

```diff
@@ -24,12 +25,50 @@ def get_num_unique_hashes(
         num_runs: int,
     ) -> int:
         random.seed(self.random_seed)
-        population = self.get_n_random_sequences(num_candidates, num_runs)
+        full_population = set()
         distinct_hashes = set()
+        last_len = -1
 
-        for i, sample in enumerate(population):
-            curr_hash = self.run_assemble_and_get_hash(file_name, i, sample)
-            distinct_hashes.add(curr_hash)
+        while len(full_population) < num_runs:
+            # It can be the case that some of the random sequences we generate
+            # are actually the same since changing AA queries can lead to fewer
+            # queries being present. Therefore, we loop here until we have enough
+            # sequences that are actually unique runs.
```

Comment:
Would it not make sense to return the new sequence whenever you try to relax queries, so that you can precisely check this? 

E.g.:

1. Generate candidate sequence
2. Run compiler and update candidate sequence, by trimming it if necessary or by extending it (if new queries appeared)
3. Use these when doing the deduplication etc

Reply:
So, you are suggesting to do the compilation as well as the trimming in `run_assemble_and_get_hash()`
I agree that I am doing too many compilations here in most cases. I will rewrite and see what you think afterward. Then we can still discuss whether we want to merge/move stuff.